### PR TITLE
feat(poloniex): createOrderWs, cancelOrderWs, cancelOrdersWs, cancelAllOrdersWs

### PR DESCRIPTION
**Pull request by samgermain**
## TS

```
% poloniex createOrderWs ADA/USDT market buy 3 0.36222 
2023-11-14T17:21:43.767Z
Node.js: v18.15.0
CCXT v4.1.50
poloniex.createOrderWs (ADA/USDT, market, buy, 3, 0.36222)
2023-11-14T17:21:49.523Z iteration 0 passed in 2759 ms

symbol |                 id | clientOrderId | timestamp | datetime | lastTradeTimestamp | type | timeInForce | postOnly | side | price | stopPrice | triggerPrice | amount | cost | average | filled | remaining | status | fee |          trades | fees | lastUpdateTimestamp | reduceOnly | takeProfitPrice | stopLossPrice
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
       | 247531504190627840 |               |           |          |                    |      |             |          |      |       |           |              |        |    0 |         |      0 |           |        |  {} | [object Object] | [{}] |                     |            |                 |              
1 objects
2023-11-14T17:21:49.523Z iteration 1 passed in 2759 ms
```

```
% poloniex cancelOrderWs '"247726801655181312"'
2023-11-15T06:21:21.497Z
Node.js: v18.15.0
CCXT v4.1.50
poloniex.cancelOrderWs (247726801655181312)
2023-11-15T06:21:26.896Z iteration 0 passed in 2605 ms

symbol |                 id | clientOrderId | timestamp | datetime | lastTradeTimestamp | type | timeInForce | postOnly | side | price | stopPrice | triggerPrice | amount | cost | average | filled | remaining | status | fee |          trades | fees | lastUpdateTimestamp | reduceOnly | takeProfitPrice | stopLossPrice
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
       | 247726801655181312 |               |           |          |                    |      |             |          |      |       |           |              |        |    0 |         |      0 |           |        |  {} | [object Object] | [{}] |                     |            |                 |              
1 objects
2023-11-15T06:21:26.896Z iteration 1 passed in 2605 ms
```

```
% poloniex cancelOrdersWs '[ "247726909205512193", "247727072754008065" ]'
2023-11-15T06:23:30.453Z
Node.js: v18.15.0
CCXT v4.1.50
poloniex.cancelOrdersWs (247726909205512193,247727072754008065)
2023-11-15T06:23:35.159Z iteration 0 passed in 2596 ms

symbol |                 id | clientOrderId | timestamp | datetime | lastTradeTimestamp | type | timeInForce | postOnly | side | price | stopPrice | triggerPrice | amount | cost | average | filled | remaining | status | fee |          trades | fees | lastUpdateTimestamp | reduceOnly | takeProfitPrice | stopLossPrice
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
       | 247727072754008065 |               |           |          |                    |      |             |          |      |       |           |              |        |    0 |         |      0 |           |        |  {} | [object Object] | [{}] |                     |            |                 |              
       | 247726909205512193 |               |           |          |                    |      |             |          |      |       |           |              |        |    0 |         |      0 |           |        |  {} | [object Object] | [{}] |                     |            |                 |              
2 objects
2023-11-15T06:23:35.159Z iteration 1 passed in 2596 ms
```

```
% poloniex cancelAllOrdersWs
2023-11-15T06:24:38.410Z
Node.js: v18.15.0
CCXT v4.1.50
poloniex.cancelAllOrdersWs ()
2023-11-15T06:24:43.154Z iteration 0 passed in 2863 ms

symbol |                 id | clientOrderId | timestamp | datetime | lastTradeTimestamp | type | timeInForce | postOnly | side | price | stopPrice | triggerPrice | amount | cost | average | filled | remaining | status | fee |          trades | fees | lastUpdateTimestamp | reduceOnly | takeProfitPrice | stopLossPrice
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
       | 247727198448922624 |               |           |          |                    |      |             |          |      |       |           |              |        |    0 |         |      0 |           |        |  {} | [object Object] | [{}] |                     |            |                 |              
       | 247727263590649856 |               |           |          |                    |      |             |          |      |       |           |              |        |    0 |         |      0 |           |        |  {} | [object Object] | [{}] |                     |            |                 |              
       | 247727369803010048 |               |           |          |                    |      |             |          |      |       |           |              |        |    0 |         |      0 |           |        |  {} | [object Object] | [{}] |                     |            |                 |              
3 objects
2023-11-15T06:24:43.154Z iteration 1 passed in 2863 ms
```

## Python

```
% py poloniex createOrderWs ADA/USDT market sell 3 0.36222
Python v3.11.6
CCXT v4.1.56
poloniex.createOrderWs(ADA/USDT,market,sell,3,0.36222)
[{'amount': None,
  'average': None,
  'clientOrderId': None,
  'cost': 0.0,
  'datetime': None,
  'fee': {'cost': None, 'currency': None, 'rate': None},
  'fees': [{'cost': None, 'currency': None, 'rate': None}],
  'filled': 0.0,
  'id': '250397211416862720',
  'info': {'clientOrderId': '',
           'code': 200,
           'message': '',
           'orderId': 250397211416862720},
  'lastTradeTimestamp': None,
  'lastUpdateTimestamp': None,
  'postOnly': None,
  'price': None,
  'reduceOnly': None,
  'remaining': None,
  'side': None,
  'status': None,
  'stopLossPrice': None,
  'stopPrice': None,
  'symbol': None,
  'takeProfitPrice': None,
  'timeInForce': None,
  'timestamp': None,
  'trades': [{'amount': None,
              'cost': None,
              'datetime': None,
              'fee': {'cost': None, 'currency': None, 'rate': None},
              'fees': [{'cost': None, 'currency': None, 'rate': None}],
              'id': None,
              'info': {'clientOrderId': '',
                       'code': 200,
                       'message': '',
                       'orderId': 250397211416862720},
              'order': '250397211416862720',
              'price': None,
              'side': None,
              'symbol': None,
              'takerOrMaker': None,
              'timestamp': None,
              'type': None}],
  'triggerPrice': None,
  'type': None}]
```

```
% py poloniex cancelOrderWs '"250398209027870720"'
Python v3.11.6
CCXT v4.1.56
poloniex.cancelOrderWs(250398209027870720)
[{'amount': None,
  'average': None,
  'clientOrderId': None,
  'cost': 0.0,
  'datetime': None,
  'fee': {'cost': None, 'currency': None, 'rate': None},
  'fees': [{'cost': None, 'currency': None, 'rate': None}],
  'filled': 0.0,
  'id': '250398209027870720',
  'info': {'clientOrderId': '',
           'code': 200,
           'message': '',
           'orderId': 250398209027870720},
  'lastTradeTimestamp': None,
  'lastUpdateTimestamp': None,
  'postOnly': None,
  'price': None,
  'reduceOnly': None,
  'remaining': None,
  'side': None,
  'status': None,
  'stopLossPrice': None,
  'stopPrice': None,
  'symbol': None,
  'takeProfitPrice': None,
  'timeInForce': None,
  'timestamp': None,
  'trades': [{'amount': None,
              'cost': None,
              'datetime': None,
              'fee': {'cost': None, 'currency': None, 'rate': None},
              'fees': [{'cost': None, 'currency': None, 'rate': None}],
              'id': None,
              'info': {'clientOrderId': '',
                       'code': 200,
                       'message': '',
                       'orderId': 250398209027870720},
              'order': '250398209027870720',
              'price': None,
              'side': None,
              'symbol': None,
              'takerOrMaker': None,
              'timestamp': None,
              'type': None}],
  'triggerPrice': None,
  'type': None}]
```

```
% py poloniex cancelOrdersWs '[ "250398209027870720", "250398235099672576" ]' 
Python v3.11.6
CCXT v4.1.56
poloniex.cancelOrdersWs(['250398209027870720', '250398235099672576'])
[{'amount': None,
  'average': None,
  'clientOrderId': None,
  'cost': 0.0,
  'datetime': None,
  'fee': {'cost': None, 'currency': None, 'rate': None},
  'fees': [{'cost': None, 'currency': None, 'rate': None}],
  'filled': 0.0,
  'id': '250398209027870720',
  'info': {'clientOrderId': '',
           'code': 21301,
           'message': 'Order not exists',
           'orderId': 250398209027870720},
  'lastTradeTimestamp': None,
  'lastUpdateTimestamp': None,
  'postOnly': None,
  'price': None,
  'reduceOnly': None,
  'remaining': None,
  'side': None,
  'status': None,
  'stopLossPrice': None,
  'stopPrice': None,
  'symbol': None,
  'takeProfitPrice': None,
  'timeInForce': None,
  'timestamp': None,
  'trades': [{'amount': None,
              'cost': None,
              'datetime': None,
              'fee': {'cost': None, 'currency': None, 'rate': None},
              'fees': [{'cost': None, 'currency': None, 'rate': None}],
              'id': None,
              'info': {'clientOrderId': '',
                       'code': 21301,
                       'message': 'Order not exists',
                       'orderId': 250398209027870720},
              'order': '250398209027870720',
              'price': None,
              'side': None,
              'symbol': None,
              'takerOrMaker': None,
              'timestamp': None,
              'type': None}],
  'triggerPrice': None,
  'type': None},
 {'amount': None,
  'average': None,
  'clientOrderId': None,
  'cost': 0.0,
  'datetime': None,
  'fee': {'cost': None, 'currency': None, 'rate': None},
  'fees': [{'cost': None, 'currency': None, 'rate': None}],
  'filled': 0.0,
  'id': '250398235099672576',
  'info': {'clientOrderId': '',
           'code': 200,
           'message': '',
           'orderId': 250398235099672576},
  'lastTradeTimestamp': None,
  'lastUpdateTimestamp': None,
  'postOnly': None,
  'price': None,
  'reduceOnly': None,
  'remaining': None,
  'side': None,
  'status': None,
  'stopLossPrice': None,
  'stopPrice': None,
  'symbol': None,
  'takeProfitPrice': None,
  'timeInForce': None,
  'timestamp': None,
  'trades': [{'amount': None,
              'cost': None,
              'datetime': None,
              'fee': {'cost': None, 'currency': None, 'rate': None},
              'fees': [{'cost': None, 'currency': None, 'rate': None}],
              'id': None,
              'info': {'clientOrderId': '',
                       'code': 200,
                       'message': '',
                       'orderId': 250398235099672576},
              'order': '250398235099672576',
              'price': None,
              'side': None,
              'symbol': None,
              'takerOrMaker': None,
              'timestamp': None,
              'type': None}],
  'triggerPrice': None,
  'type': None}]
```

```
% py poloniex cancelAllOrdersWs
Python v3.11.6
CCXT v4.1.56
poloniex.cancelAllOrdersWs()
[{'amount': None,
  'average': None,
  'clientOrderId': None,
  'cost': 0.0,
  'datetime': None,
  'fee': {'cost': None, 'currency': None, 'rate': None},
  'fees': [{'cost': None, 'currency': None, 'rate': None}],
  'filled': 0.0,
  'id': '250399316147961856',
  'info': {'clientOrderId': '',
           'code': 200,
           'message': '',
           'orderId': 250399316147961856},
  'lastTradeTimestamp': None,
  'lastUpdateTimestamp': None,
  'postOnly': None,
  'price': None,
  'reduceOnly': None,
  'remaining': None,
  'side': None,
  'status': None,
  'stopLossPrice': None,
  'stopPrice': None,
  'symbol': None,
  'takeProfitPrice': None,
  'timeInForce': None,
  'timestamp': None,
  'trades': [{'amount': None,
              'cost': None,
              'datetime': None,
              'fee': {'cost': None, 'currency': None, 'rate': None},
              'fees': [{'cost': None, 'currency': None, 'rate': None}],
              'id': None,
              'info': {'clientOrderId': '',
                       'code': 200,
                       'message': '',
                       'orderId': 250399316147961856},
              'order': '250399316147961856',
              'price': None,
              'side': None,
              'symbol': None,
              'takerOrMaker': None,
              'timestamp': None,
              'type': None}],
  'triggerPrice': None,
  'type': None},
 {'amount': None,
  'average': None,
  'clientOrderId': None,
  'cost': 0.0,
  'datetime': None,
  'fee': {'cost': None, 'currency': None, 'rate': None},
  'fees': [{'cost': None, 'currency': None, 'rate': None}],
  'filled': 0.0,
  'id': '250399348599291904',
  'info': {'clientOrderId': '',
           'code': 200,
           'message': '',
           'orderId': 250399348599291904},
  'lastTradeTimestamp': None,
  'lastUpdateTimestamp': None,
  'postOnly': None,
  'price': None,
  'reduceOnly': None,
  'remaining': None,
  'side': None,
  'status': None,
  'stopLossPrice': None,
  'stopPrice': None,
  'symbol': None,
  'takeProfitPrice': None,
  'timeInForce': None,
  'timestamp': None,
  'trades': [{'amount': None,
              'cost': None,
              'datetime': None,
              'fee': {'cost': None, 'currency': None, 'rate': None},
              'fees': [{'cost': None, 'currency': None, 'rate': None}],
              'id': None,
              'info': {'clientOrderId': '',
                       'code': 200,
                       'message': '',
                       'orderId': 250399348599291904},
              'order': '250399348599291904',
              'price': None,
              'side': None,
              'symbol': None,
              'takerOrMaker': None,
              'timestamp': None,
              'type': None}],
  'triggerPrice': None,
  'type': None}]
```

### CODE REVIEW

**Comment by samgermain** [comment]
The handler currently doesn't work in python

**Comment by carlosmiei** [comment]
> The handler currently doesn't work in python

what do you mean exactly by this?

**Comment by samgermain** [comment]
> > The handler currently doesn't work in python
> 
> what do you mean exactly by this?

I dug deeper and it looks like the `client.resolve` method doesn't return anything, but only in python. I can put a `print(orders)` statement right before `client.resolve` and the expected value of orders is printed though: this is not the case in Typescript.

**Comment by samgermain** [comment]
> > The handler currently doesn't work in python
> 
> what do you mean exactly by this?

It's really weird behaviour, you can see that the message hash matches with the print statements, and the typescript file resolves properly, 

```
% poloniex createOrderWs ADA/USDT limit buy 4 0.3    
2023-11-20T09:20:51.654Z
Node.js: v18.15.0
CCXT v4.1.56
poloniex.createOrderWs (ADA/USDT, limit, buy, 4, 0.3)
2023-11-20T09:20:54.040Z iteration 0 passed in 1115 ms

symbol |                 id | clientOrderId | timestamp | datetime | lastTradeTimestamp | type | timeInForce | postOnly | side | price | stopPrice | triggerPrice | amount | cost | average | filled | remaining | status | fee |          trades | fees | lastUpdateTimestamp | reduceOnly | takeProfitPrice | stopLossPrice
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
       | 249584803408523264 |               |           |          |                    |      |             |          |      |       |           |              |        |    0 |         |      0 |           |        |  {} | [object Object] | [{}] |                     |            |                 |              
1 objects
2023-11-20T09:20:54.040Z iteration 1 passed in 1115 ms
```

but the python file (with print statements) doesn't

```
% py poloniex createOrderWs ADA/USDT limit buy 4 0.3 
Python v3.11.6
CCXT v4.1.56
poloniex.createOrderWs(ADA/USDT,limit,buy,4,0.3)
None
1700471808304
1700471808304
None
None
```

File: ts/src/pro/poloniex.ts [hunk 1]
```diff
@@ -155,6 +164,160 @@ export default class poloniex extends poloniexRest {
...
+            uppercaseType = 'LIMIT_MAKER';
+        }
+        const request = {
+            'symbol': market['id'],
+            'side': side.toUpperCase (),
+            'type': type.toUpperCase (),
+        };
+        if ((uppercaseType === 'MARKET') && (uppercaseSide === 'BUY')) {
+            let quoteAmount = this.safeString (params, 'amount');
+            if (quoteAmount === undefined) {
```
Comment (carlosmiei) [inline ts/src/pro/poloniex.ts:225]: @samgermain I think we should use the createMarketBuyRequiresPrice logic here

**Comment by carlosmiei** [comment]
DEMO

```
p poloniex createOrderWs "LTC/USDT" market buy 1 5
Python v3.11.5
CCXT v4.1.56
poloniex.createOrderWs(LTC/USDT,market,buy,1,5)
[{'amount': None,
  'average': None,
  'clientOrderId': None,
  'cost': 0.0,
  'datetime': None,
  'fee': {'cost': None, 'currency': None, 'rate': None},
  'fees': [{'cost': None, 'currency': None, 'rate': None}],
  'filled': 0.0,
  'id': '250698770256232453',
  'info': {'clientOrderId': '',
           'code': 200,
           'message': '',
           'orderId': 250698770256232453},
  'lastTradeTimestamp': None,
  'lastUpdateTimestamp': None,
  'postOnly': None,
  'price': None,
  'reduceOnly': None,
  'remaining': None,
  'side': None,
  'status': None,
  'stopLossPrice': None,
  'stopPrice': None,
  'symbol': None,
  'takeProfitPrice': None,
  'timeInForce': None,
  'timestamp': None,
  'trades': [{'amount': None,
              'cost': None,
              'datetime': None,
              'fee': {'cost': None, 'currency': None, 'rate': None},
              'fees': [{'cost': None, 'currency': None, 'rate': None}],
              'id': None,
              'info': {'clientOrderId': '',
                       'code': 200,
                       'message': '',
                       'orderId': 250698770256232453},
              'order': '250698770256232453',
              'price': None,
              'side': None,
              'symbol': None,
              'takerOrMaker': None,
              'timestamp': None,
              'type': None}],
  'triggerPrice': None,
  'type': None}]
```
```
p poloniex createOrderWs "LTC/USDT" limit buy 0.1 50
Python v3.11.5
CCXT v4.1.56
poloniex.createOrderWs(LTC/USDT,limit,buy,0.1,50)
[{'amount': None,
  'average': None,
  'clientOrderId': None,
  'cost': 0.0,
  'datetime': None,
  'fee': {'cost': None, 'currency': None, 'rate': None},
  'fees': [{'cost': None, 'currency': None, 'rate': None}],
  'filled': 0.0,
  'id': '250699088360640512',
  'info': {'clientOrderId': '',
           'code': 200,
           'message': '',
           'orderId': 250699088360640512},
  'lastTradeTimestamp': None,
  'lastUpdateTimestamp': None,
  'postOnly': None,
  'price': None,
  'reduceOnly': None,
  'remaining': None,
  'side': None,
  'status': None,
  'stopLossPrice': None,
  'stopPrice': None,
  'symbol': None,
  'takeProfitPrice': None,
  'timeInForce': None,
  'timestamp': None,
  'trades': [{'amount': None,
              'cost': None,
              'datetime': None,
              'fee': {'cost': None, 'currency': None, 'rate': None},
              'fees': [{'cost': None, 'currency': None, 'rate': None}],
              'id': None,
              'info': {'clientOrderId': '',
                       'code': 200,
                       'message': '',
                       'orderId': 250699088360640512},
              'order': '250699088360640512',
              'price': None,
              'side': None,
              'symbol': None,
              'takerOrMaker': None,
              'timestamp': None,
              'type': None}],
  'triggerPrice': None,
  'type': None}]
```
```
 p poloniex cancelOrderWs "250699088360640512"                         
Python v3.11.5
CCXT v4.1.56
poloniex.cancelOrderWs(250699088360640512)
[{'amount': None,
  'average': None,
  'clientOrderId': None,
  'cost': 0.0,
  'datetime': None,
  'fee': {'cost': None, 'currency': None, 'rate': None},
  'fees': [{'cost': None, 'currency': None, 'rate': None}],
  'filled': 0.0,
  'id': '250699088360640512',
  'info': {'clientOrderId': '',
           'code': 200,
           'message': '',
           'orderId': 250699088360640512},
  'lastTradeTimestamp': None,
  'lastUpdateTimestamp': None,
  'postOnly': None,
  'price': None,
  'reduceOnly': None,
  'remaining': None,
  'side': None,
  'status': None,
  'stopLossPrice': None,
  'stopPrice': None,
  'symbol': None,
  'takeProfitPrice': None,
  'timeInForce': None,
  'timestamp': None,
  'trades': [{'amount': None,
              'cost': None,
              'datetime': None,
              'fee': {'cost': None, 'currency': None, 'rate': None},
              'fees': [{'cost': None, 'currency': None, 'rate': None}],
              'id': None,
              'info': {'clientOrderId': '',
                       'code': 200,
                       'message': '',
                       'orderId': 250699088360640512},
              'order': '250699088360640512',
              'price': None,
              'side': None,
              'symbol': None,
              'takerOrMaker': None,
              'timestamp': None,
              'type': None}],
  'triggerPrice': None,
  'type': None}]
```

**Pull request merged**